[WIP] Remove disabled products from cart

rust-le opened the pull request:
| Q               | A
|-----------------|-----
| Branch?         | 1.14
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | https://github.com/Sylius/Sylius/issues/15933
| License         | MIT

<!--
 - Bug fixes must be submitted against the 1.14 or 2.0 branch
 - Features and deprecations must be submitted against the 2.1 branch
 - Make sure that the correct base branch is set

 To be sure you are not breaking any Backward Compatibilities, check the documentation:
 https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html
-->

rust-le committed:
Add subscriber

rust-le committed:
Add simple Behat test

rust-le committed:
Add simple Behat test

# Conflicts:
#	features/cart/shopping_cart/remove_product_from_cart_when_not_available.feature

rust-le committed:
Add flashes in all languages

Rafikooo commented on src/Sylius/Bundle/ShopBundle/EventSubscriber/EnabledOnlyProductsInCartSubscriber.php:
```diff
@@ -0,0 +1,116 @@
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+final class EnabledOnlyProductsInCartSubscriber implements EventSubscriberInterface
```
1. We have a more appropriate entry point for cart-related processing. I’d suggest introducing a `RemoveUnavailableCartItemsProcessor` service implementing `Sylius\Component\Order\Processor\OrderProcessorInterface`.
2. This approach makes the logic easier to test and integrates more naturally with the existing Sylius order recalculation flow.
3. Since the availability check is relatively heavy, it would be cleaner to move it into a dedicated service, e.g. `ProductVariantChannelEligibilityChecker`, and call it from the processor.

rust-le replied:
done

rust-le replied:
done

Rafikooo commented on features/cart/shopping_cart/remove_product_from_cart_when_not_available.feature:
```diff
@@ -0,0 +1,55 @@
+@shopping_cart
+Feature: Remove product from cart when not available
+    In order to avoid checkout of unavailable products
+    As a shop user
```
```suggestion
    As a Customer
```

rust-le replied:
done

Rafikooo commented on features/cart/shopping_cart/remove_product_from_cart_when_not_available.feature:
```diff
@@ -0,0 +1,55 @@
+
+    @ui
+    Scenario: Removing product from cart when it is disabled
+        Given I am a logged in customer
```
```suggestion
        Given I am a logged in customer
        And the store has a product "PHP T-Shirt" priced at "$10.99"
        And the store has a product "PHP Mug" priced at "$5.99"
        /\ Move the steps above to Background
        When I add  "PHP T-Shirt" to the cart
        And I add "PHP Mug" to the cart
```

rust-le replied:
done

Rafikooo commented on features/cart/shopping_cart/remove_product_from_cart_when_not_available.feature:
```diff
@@ -0,0 +1,55 @@
+        And I add this product to the cart
+        And the store has a product "PHP Mug" priced at "$5.99"
+        And I add this product to the cart
+        And the "PHP Mug" product is disabled
```
```suggestion
        And the "PHP Mug" product has been disabled
```

rust-le replied:
done

Rafikooo commented on features/cart/shopping_cart/remove_product_from_cart_when_not_available.feature:
```diff
@@ -0,0 +1,55 @@
+    Background:
+        Given the store operates on a single channel in "United States"
+
+    @ui
```
```suggestion
    @api @ui
```

These scenarios apply to both UI and API. However, it would be good to add one `@no-api @ui` scenario that also checks the presence of the flash message.

rust-le replied:
after changes works with @api

Rafikooo commented on src/Sylius/Bundle/ShopBundle/Resources/translations/flashes.ar.yml:
```diff
@@ -0,0 +1,4 @@
+sylius:
+    cart:
+        product:
+            not_available: 'بعض المنتجات غير متوفرة وتم إزالتها من السلة.'
```
Please add a trailing newline at the end of this and other `flash.*.yml` files.  
Also, rename their extensions from `.yml` to `.yaml` as we use `.yaml` across the project.

rust-le replied:
done

rust-le committed:
Rewrited to use sylius.order_processing

rust-le committed:
Rewrited to use sylius.order_processing

# Conflicts:
#	src/Sylius/Component/Core/OrderProcessing/Subscriber/CartItemsRemovedFlashSubscriber.php

rust-le committed:
Fixed typo

rust-le committed:
Fixed trnaslations, add recalculation on sylius.cart_summary

rust-le committed:
Behat test pass in @ui and @api

rust-le committed:
Add session existence check

rust-le committed:
Changes in PHPUnit tests and test fixtures

rust-le committed:
Add snippet for suite step

Rafikooo reviewed (commented):
## Performance Concern: Running processor on every cart view

  Currently, the processor runs on **every** `cart_summary` event (every cart page view), which means:
  - Full order processing chain executes on each page refresh
  - For a cart with 50 items, the eligibility checker runs 50 times per view
  - This happens even when products haven't changed

  **Suggested optimization:**

  Use Doctrine `preUpdate` event instead:

  1. **When admin disables a product** → mark affected carts in cache
  2. **On `cart_summary`** → quick cache check (< 1ms)
  3. **If flagged** → run processor + clear flag

  **Benefits:**
  - 99% of cart views = lightweight cache check only
  - User still gets "live" feedback on next page load
  - Significant performance improvement for large carts

  This maintains the desired UX while avoiding unnecessary processing overhead.

rust-le commented:
outdated, closing
